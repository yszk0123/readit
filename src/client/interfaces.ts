import { Action as ReduxAction } from 'redux';

// Type

export type ID = string;

export type DateOnly = string;

export enum ReviewStatus {
  PLAN_TO_BUY = 'PLAN_TO_BUY',
  UNREAD = 'UNREAD',
  READING = 'READING',
  READ = 'READ',
  STOP = 'STOP',
}

export type ReviewRating = 0 | 1 | 2 | 3 | 4 | 5;

// Entity

export interface User {
  id: ID;
  nickname: string;
  email: string | null;
}

export interface Book {
  id: ID;
  title: string;
  subtitle: string | null;
  author: string;
  thumbnailLink: string;
  description: string;

  category: string;
  pageCount: number;
  publisher: string;
  publishedAt: DateOnly;
  isbn: string | null;
  meta: {
    source: {
      type: string;
      link: string;
      searchText: string;
    };
  };
}

export interface Review {
  id: ID;
  status: ReviewStatus;
  rating: ReviewRating;
}

export interface ReadingLog {
  id: ID;
  book: Book;
  review: Review;
  removed?: boolean;
}

// Action

export interface Action extends ReduxAction {
  payload: any;
}

export enum ActionTypes {
  HELLO = 'HELLO',
  CHANGE_NAME = 'CHANGE_NAME',
  FETCH_SUCCESS = 'FETCH_SUCCESS',
  FETCH_FAILURE = 'FETCH_FAILURE',
  FETCH_READING_LOGS = 'FETCH_READING_LOGS',
  CREATE_READING_LOG = 'CREATE_READING_LOG',
  UPDATE_REVIEW = 'UPDATE_REVIEW',
  SEARCH_BOOK_BY_TITLE = 'SEARCH_BOOK_BY_TITLE',
  REMOVE_READING_LOG = 'REMOVE_READING_LOG',
  REMOVE_READING_LOG_SUCCESS = 'REMOVE_READING_LOG_SUCCESS',
  REMOVE_ENTITY = 'REMOVE_ENTITY',
}

// State

export interface State {
  entities: {
    users: Record<ID, User>;
    books: Record<ID, Book>;
    reviews: Record<ID, Review>;
    readingLogs: Record<ID, ReadingLog>;
  };
  ui: {
    currentUser: ID | null;
    readingLogs: ID[];
    titleInput: string;
  };
}
