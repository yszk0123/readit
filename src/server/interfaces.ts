export interface IModels {
  User: any;
  Book: any;
  Review: any;
}
