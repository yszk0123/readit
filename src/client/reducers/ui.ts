import { Action, ActionTypes, State } from '../interfaces';
import merge from '../utils/merge';

const initialState: State['ui'] = {
  currentUser: null,
  readingLogs: [],
  books: [],
  reviews: [],
  titleInput: '',
};

export default function ui(state: State['ui'] = initialState, action: Action) {
  switch (action.type) {
    case ActionTypes.FETCH_SUCCESS:
      return merge(state, action.payload.ui);
    case ActionTypes.SEARCH_BOOK_BY_TITLE:
      return { ...state, titleInput: action.payload.title };
    default:
      return state;
  }
}
