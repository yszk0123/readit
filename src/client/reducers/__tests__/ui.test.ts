import { ActionTypes } from '../../interfaces';
import { ID } from '../../interfaces';
import ui from '../ui';

describe('ui', () => {
  function fetchSuccess(reviewId: ID) {
    return {
      type: ActionTypes.FETCH_SUCCESS,
      payload: { ui: { reviews: [reviewId] } },
    };
  }
  const noopAction = { type: 'NOOP', payload: {} };

  it('should merge payload when fetch success', () => {
    expect(ui(undefined, fetchSuccess('123'))).toMatchSnapshot();
  });

  it('should not add duplicate id when fetch success', () => {
    const reviewId = '123';
    const nextState = ui(undefined, fetchSuccess(reviewId));
    expect(ui(nextState, fetchSuccess(reviewId))).toMatchSnapshot();
  });

  it('should add second id when fetch success', () => {
    const initialState = ui(undefined, noopAction);
    const nextState = ui(initialState, fetchSuccess('123'));
    expect(ui(nextState, fetchSuccess('234'))).toMatchSnapshot();
  });
});
