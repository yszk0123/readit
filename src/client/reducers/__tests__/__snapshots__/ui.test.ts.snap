// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui should add second id when fetch success 1`] = `
Object {
  "currentUser": null,
  "readingLogs": Array [],
  "reviews": Array [
    "123",
    "234",
  ],
  "titleInput": "",
}
`;

exports[`ui should merge payload when fetch success 1`] = `
Object {
  "currentUser": null,
  "readingLogs": Array [],
  "reviews": Array [
    "123",
  ],
  "titleInput": "",
}
`;

exports[`ui should not add duplicate id when fetch success 1`] = `
Object {
  "currentUser": null,
  "readingLogs": Array [],
  "reviews": Array [
    "123",
  ],
  "titleInput": "",
}
`;
